### PR TITLE
feat: add multi-stages docker and support platform arm

### PR DESCRIPTION
**Pull request by phv2312**
## Description

- Support platform linux/arm64
- Support multiple stages (lite & full) version of Dockerfile. With full version will install the `unstructured[all-docs]`
- When release event triggered. It will update: `{semver}-lite`, `{semver}-full`, `latest-lite`, `latest-full` images. Each image has platform `linux/arm64` & `linux/amd64`

This PR is inherited from PR https://github.com/Cinnamon/kotaemon/pull/165 (for multi-stages idead) & https://github.com/Cinnamon/kotaemon/pull/226 (for platform arm64) and @cin-niko . Thank you guys for your great contribution.

## Type of change

- [x] New features (non-breaking change).
- [ ] Bug fix (non-breaking change).
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected).

## Checklist

- [x] I have performed a self-review of my code.
- [ ] I have added thorough tests if it is a core feature.
- [ ] There is a reference to the original bug report and related work.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [ ] The feature is well documented.

### CODE REVIEW

**Comment by cin-niko** [comment]
As #254 , Should we add `fastembed` to `kotaemon` dependencies a @phv2312 ?

**Comment by taprosoft** [comment]
@phv2312 also please refer here https://github.com/Cinnamon/kotaemon/blob/cd85c4935c85afdd18416fd902e09c4a8b997d86/libs/kotaemon/tests/conftest.py#L63 for list of packages that are currently skipped.
Please include:

* `sentence_trasnformers`
* `llama-cpp-python`
* `fastembed`

**Comment by phv2312** [comment]
@taprosoft @cin-niko should we streamline & add those packages in the pyproject.toml (for example: kotaemon/pyproject.toml). I think we can add additional mode, called `full` (or another name) and list out some required dependencies to experience full features.
Example: https://github.com/Cinnamon/kotaemon/blob/main/libs/kotaemon/pyproject.toml#L74

**Comment by taprosoft** [comment]
@phv2312 yes let update it in the `pyproject.toml` add an additional extra.

**Comment by cin-niko** [comment]
I think we should discuss about how to manage kotaemon dependencies.
If we add new dependencies (vectordb, embedding, new rag) to `pyproject.toml`. It will bloat our dependencies in the near future.
- Option 1 (as a @phv2312): Keep a base mode and add new deps to `full` (or `adv`) optional-dependencies.
- Option 2 ([current](https://github.com/Cinnamon/kotaemon/blob/cd85c4935c85afdd18416fd902e09c4a8b997d86/libs/kotaemon/kotaemon/loaders/utils/pdf_ocr.py#L35) ): Raise ImportError and ask users for installing (auto install or ask in the UI ?).

**Comment by phv2312** [comment]
Ask user for installing is quite hard especially when we're inside Docker. We may need to support some functions, such as `_try_import_or_download()`

**Comment by cin-niko** [comment]
I've found the same idea [here](https://github.com/ultralytics/ultralytics/blob/1210b49cd2cb21bb70ed92b04f872317c18d1fbb/ultralytics/utils/checks.py#L348)
```py
def check_requires(..., install = True):
    ...
```

**Comment by phv2312** [comment]
Good points @cin-niko . How about some complicated which require apt install?

**Comment by RicardoMonteiroSimoes** [comment]
Couldn't this be done in a simple matrix, that produces _n_ images with varying dependencies? The setup to get it to going is a bit harder, but it will offer full flexibility for all use cases.

**Comment by phv2312** [comment]
@RicardoMonteiroSimoes 
Oh, you means that we should prepare different Dockerfile for each mode (Dockerfile.lite/ Dockerfile.full as example) ? Then later if we have different mode, just come up with different Dockerfile is okay ?

**Comment by RicardoMonteiroSimoes** [comment]
> @RicardoMonteiroSimoes Oh, you means that we should prepare different Dockerfile for each mode (Dockerfile.lite/ Dockerfile.full as example) ? Then later if we have different mode, just come up with different Dockerfile is okay ?

Basically, yeah. Normally when dealing with dependencies people can define them as "supplied" or "included", or specifically ask to exclude them when importing (i.e. maven in Java). Considering this would be rather hard here, why not decide on 3-5 versions and just have them built for amd and arm platforms?

**Comment by phv2312** [comment]
I see, thank you for your suggestion. Let me discuss with our team first, Btw I think go with your suggestion will have more room for the flexibility later. 
But I will consider it as enhancement, and intend to accomplish it in another PR (or if you or somebody can help me it's more than welcome), I want to finish this PR soon so that everyone can test the new docker for platform arm & got feedback first.

**Comment by RicardoMonteiroSimoes** [comment]
Sure thing! I did work on this here which is open https://github.com/Cinnamon/kotaemon/pull/226 and introduces ARM support in the docker build workflow. I also tested it partially locally, and it seems to run.

**Comment by cin-niko** [comment]
> How about some complicated which require apt install?

@phv2312, IMO 
- Without docker: We need a clear document about pre-requisites for each feature. 
- With docker: We can install all dependencies in the full version

Anw, this PR is for ARM docker. We can solve the dependencies management in a future PR.

**Comment by lone17** [comment]
> @taprosoft @cin-niko should we streamline & add those packages in the pyproject.toml (for example: kotaemon/pyproject.toml). I think we can add additional mode, called `full` (or another name) and list out some required dependencies to experience full features. Example: https://github.com/Cinnamon/kotaemon/blob/main/libs/kotaemon/pyproject.toml#L74

+1 on this, this is what we intended when we set up the `toml` file. However, this requires extra care when adding new features: decide whether it should be included by default; if not then implement a fallback mechanism whenever users attempt to use not-installed features.

Ideally, we would like to build an extension-based framework for installing optional feature #48. But that's just an idea then.

**Comment by RicardoMonteiroSimoes** [comment]
Small input from my side:

I would recommend 3 baseline versions:
- image-local (only using ollama)
- image-online (only using AI services)
- image-full (using both) <-- dont know how applicable that is?

Then, you'd have baselines like:
- api only
- FE only
- full document coverage (for example, saw a different PR talking about parsing documents)

Of course this would already be 9 different images. Another option, which is a bit harder for non-developers, would be docker compose. Basically 3/4 images:
- main AI one (includes API)
  - local
  - online
- FE one
- parser one

This means people could just "start" what they need, or use `docker-compose` to run it all at once.
(Docker compose can trigger a download of files if necessary, so it shouldn't be _that_ hard)

File: README.md [hunk 1]
```diff
@@ -86,17 +86,44 @@ Use the most recent release .zip to include latest features and bug-fixes.
...
+```
+
+- To use the `lite` version.
+
+```
+docker run \
+-e GRADIO_SERVER_NAME=0.0.0.0 \
+-e GRADIO_SERVER_PORT=7860 \
+-p 7860:7860 -it --rm \
+ghcr.io/cinnamon/kotaemon:latest-full
```
Comment (taprosoft) [inline README.md:109]: latest-lite

Comment (taprosoft) [reply README.md:109]: Also please move lite image to first for most of mainstream users

File: README.md [hunk 2]
```diff
@@ -86,17 +86,44 @@ Use the most recent release .zip to include latest features and bug-fixes.
...
+```
+
+Then, navigate to `http://localhost:7860/` to access the web UI.
+
+Currently, two platforms: `linux/amd64` and `linux/arm64` (for newer Mac) are provided & tested. User can specify the platform by passing `--platform` in the docker run command. For example:
```
Comment (taprosoft) [inline README.md:112]: Move this before navigate section

**Review by taprosoft — APPROVED** [review]
LGTM

**Pull request merged**